nifi-nar-bundles/nifi-kudu-bundle/nifi-kudu-processors/src/main/java/org/apache/nifi/processors/kudu/ScanKudu.java

```diff
@@ -0,0 +1,398 @@
+                throw new RuntimeException(e);
+            }
+
+            rowsPulled++;
+
+            // bulkSize controls number of records per flow file.
+            if (batchSize > 0 && rowsPulled >= batchSize) {
+                finalizeFlowFile(session, flowFile, tableName, rowsPulled, null);
+                flowFile = null;
+                rowsPulledHolder.set(0);
+            } else {
+                rowsPulledHolder.set(rowsPulled);
+            }
+        }
+
+        public boolean isHandledAny(){
+            return handledAny;
+        }
+
+        @Override
+        public FlowFile getFlowFile(){
+            return flowFile;
+        }
+
+        public long getRecordsCount(){
+            return rowsPulledHolder.get();
+        }
+
+    }
+
+
+    /*
+     * Initiates FF content, adds relevant attributes, and starts content with JSON array "["
+     */
+    private FlowFile initNewFlowFile(final ProcessSession session, final FlowFile origFF, final String tableName) throws IOException{
+
+        FlowFile flowFile = session.create(origFF);
+        flowFile = session.putAttribute(flowFile, KUDU_TABLE_ATTR, tableName);
+        flowFile = session.putAttribute(flowFile, CoreAttributes.MIME_TYPE.key(), "application/json");
+
+        final AtomicReference<IOException> ioe = new AtomicReference<>(null);
+        flowFile = session.write(flowFile, (out) -> {
+            try{
+                out.write("[".getBytes());
+            }catch(IOException e){
+                ioe.set(e);
+            }
+        });
+
+        if (ioe.get() != null){
+            throw ioe.get();
+        }
+
+        return flowFile;
+    }
+
+    private void finalizeFlowFile(final ProcessSession session, FlowFile flowFile, final String tableName,
+                                  Long rowsPulled, Exception e) {
+        Relationship rel = REL_SUCCESS;
+        flowFile = session.putAttribute(flowFile, KUDU_ROWS_COUNT_ATTR, rowsPulled.toString());
+
+        final AtomicReference<IOException> ioe = new AtomicReference<>(null);
+        flowFile = session.append(flowFile, (out) -> {
+            try{
+                out.write("]".getBytes());
+            } catch(IOException ei) {
+                ioe.set(ei);
+            }
+        });
+        if (e != null || ioe.get() != null) {
+            flowFile = session.putAttribute(flowFile, "scankudu.error", (e==null?e:ioe.get()).toString());
+            rel = REL_FAILURE;
+        } else {
+            session.getProvenanceReporter().receive(flowFile, tableName);
+        }
+        session.transfer(flowFile, rel);
+    }
+
+    private void addPredicate(KuduScanner.KuduScannerBuilder scannerBuilder, KuduTable kuduTable, String column, Object value, KuduPredicate.ComparisonOp comparisonOp) {
+        ColumnSchema columnSchema = kuduTable.getSchema().getColumn(column);
+        KuduPredicate predicate = KuduPredicate.newComparisonPredicate(columnSchema, comparisonOp, value);
```

Comment:
`KuduPredicate.newComparisonPredicate` will throw an exception if the column contains non-string values (because the `value` parameter is actually a `String`).

The `value` needs to be parsed based on the column type.

This method should be changed to something like this:
```java
    private void addPredicate(KuduScanner.KuduScannerBuilder scannerBuilder, KuduTable kuduTable, String column, String value, KuduPredicate.ComparisonOp comparisonOp) {
        ColumnSchema columnSchema = kuduTable.getSchema().getColumn(column);
        Object parsedValue = parseValue(value, columnSchema);
        KuduPredicate predicate = KuduPredicate.newComparisonPredicate(columnSchema, comparisonOp, parsedValue);
        scannerBuilder.addPredicate(predicate);
    }
```

And the `parseValue` method should be created and probably added to `AbstractKuduProcessor`.
It would look something like this, I added some implementations but not all, left TODOs:
```java
    protected Object parseValue(String value, ColumnSchema columnSchema) {
        Object  parsedValue;

        Type type = columnSchema.getType();

        switch (type) {
            case STRING:
                parsedValue = value;
                break;
            case INT8:
                // TODO
                parsedValue = value;
                break;
            case INT16:
                // TODO
                parsedValue = value;
                break;
            case INT32:
                parsedValue = Integer.valueOf(value);
                break;
            case INT64:
                parsedValue = Long.valueOf(value);
                break;
            case BOOL:
                parsedValue = Boolean.valueOf(value);
                break;
            case DECIMAL:
                // TODO
                parsedValue = value;
                break;
            case FLOAT:
                parsedValue = Float.valueOf(value);
                break;
            case DOUBLE:
                // TODO
                parsedValue = value;
                break;
            case UNIXTIME_MICROS:
                // TODO
                parsedValue = value;
                break;
            case BINARY:
                // TODO
                parsedValue = value;
                break;
            default:
                // TODO Need to handle this
                throw new IllegalArgumentException("Couldn't parse '" + value + "' as '" + type + "'");
        }

        return parsedValue;
    }
```

Reply:
https://github.com/apache/nifi/pull/3387/files/f2ea2ddfec9929f7baa9c41369445591310a36d4#r303540283
I had similar logic but I was confused here, sorry.